Miglioramenti al plugin Hacker News

gianluca-nitti opened the pull request:
Ora notifica le prime 5 top stories ogni 6 ore.

gianluca-nitti committed:
Miglioramenti al plugin Hacker News

Ora notifica le prime 5 top stories ogni 6 ore.

doppioandante commented on src/main/java/com/neroreflex/plugins/HackerNews.java:
```diff
@@ -31,64 +31,62 @@
 public final class HackerNews extends Trancio {
 
-    private int interval = 30000; //tempo tra le chiamate all'api in millisecondi
+    private int interval = (int)2.16e+7; // (= 6h) tempo tra le chiamate all'api in millisecondi
```
Conviene usare `Duration.ofHours(6).toMillis()`
In un int ci starebbe, ma siccome `toMillis` ritorna `long` tanto vale cambiare il tipo di interval.

doppioandante commented on src/main/java/com/neroreflex/plugins/HackerNews.java:
```diff
@@ -31,64 +31,62 @@
+                url = obj.getString("url");
+            }catch(NullPointerException e){
+                url = defaultURL + id;
+            }
```
Forse si può usare il metodo `putIfAbsent` ereditato da Map invece di usare un try / catch, ma va bene anche così.

doppioandante commented on src/main/java/com/neroreflex/plugins/HackerNews.java:
```diff
@@ -31,64 +31,62 @@
+            return other.score - score;
+        }
+        public String toString(){
+            return "From Hacker News: " + title + " " + url + " (score = " + score + ")";
```
Forse si può togliere lo score, in alternativa si può lasciare solamente il numero tra parentesi, segnalare che si tratta dello score non è molto importante e aggiunge caratteri (su IRC lo spazio è prezioso)

doppioandante commented:
Qualche commento più stilistico che altro, per me può andare.

gianluca-nitti committed:
Fixati alcuni dettagli

gianluca-nitti commented:
@doppioandante ho sistemato le 2 cose che hai detto. Però ho notato che:
- filtrare per punteggio non è la cosa migliore perchè nelle top 500 le migliori per punteggio possono essere vecchie, ad esempio stamattina le prime 2 (per score) sono ancora l'hack di dropbox e il razzo esploso, che sono di ieri. Cosa che non avviena sulla home di HN.
- la home evidentemente usa un qualche algoritmo che le ordina in base al punteggio **e** al quanto sono nuove.
- gli id restituiti dall'endpoint delle top 500 sembrano essere ordinati proprio come le notizie sulla home (ho controllato a mano i primi 10 e corrispondono).

Comincio a pensare che forse mettere direttamente le prime restituite dall'API senza ordinarle per punteggio sia meglio.

doppioandante commented:
Sì sono d'accordo

gianluca-nitti committed:
Ora usa l'ordinamento fornito dall'API.

Non sono ordinate per punteggio ma sono le prime 5 della home.
Il bot invece che ordinare manualmente le prime 5 news per punteggio si
affida all'algoritmo di ordinamento del server.

(in questo modo il codice si semplifica anche molto... KISS intensifies)

gianluca-nitti commented:
Fatto.

doppioandante commented on src/main/java/com/neroreflex/plugins/HackerNews.java:
```diff
@@ -31,64 +32,43 @@
+            JsonArray apiResponse = (JsonArray)jsonApiCall(apiTopStories); //ottiene l'elenco degli id delle top stories, max 500 (quelle in home)
+            for(int i = 0; i < 5; i++){
+                int id = apiResponse.getInt(i, -1); //id della i-esima notizia, o -1 se per qualche motivo la risposta dell'api fosse più corta
+                if(i == -1) break; //esce se l'id non è stato trovato
```
Forse è meglio un `continue` qui, magari qualcosa nel ciclo dopo si raccimola
